File: llamafile/server/www/highlight_ld.js
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

const LD_KEYWORDS = new Set([
  'AFTER',
  'ASSERT',
  'AS_NEEDED',
  'AT',
  'BEFORE',
  'BYTE',
  'COMMON',
  'CONSTRUCTORS',
  'COPY',
  'CREATE_OBJECT_SYMBOLS',
  'DSECT',
  'ENTRY',
  'EXTERN',
  'FILEHDR',
  'FILL',
  'FLAGS',
  'FORCE_COMMON_ALLOCATION',
  'GROUP',
  'HIDDEN',
  'INCLUDE',
  'INFO',
  'INHIBIT_COMMON_ALLOCATION',
  'INPUT',
  'INSERT',
  'KEEP',
  'LD_FEATURE',
  'LONG',
  'MEMORY',
  'NOCROSSREFS',
  'NOCROSSREFS_TO',
  'NOLOAD',
  'ONLY_IF_RO',
  'ONLY_IF_RW',
  'OUTPUT',
  'OUTPUT_ARCH',
  'OUTPUT_FORMAT',
  'OVERLAY',
  'PHDR',
  'PHDRS',
  'PROVIDE',
  'PROVIDE_HIDDEN',
  'PT_DYNAMIC',
  'PT_GNU_STACK',
  'PT_INTERP',
  'PT_LOAD',
  'PT_NOTE',
  'PT_NULL',
  'PT_PHDR',
  'PT_SHLIB',
  'PT_TLS',
  'QUAD',
  'REGION_ALIAS',
  'SEARCH_DIR',
  'SECTIONS',
  'SHORT',
  'SORT',
  'SORT_BY_ALIGNMENT',
  'SORT_BY_INIT_PRIORITY',
  'SORT_BY_NAME',
  'SORT_NONE',
  'SQUAD',
  'STARTUP',
  'SUBALIGN',
  'TARGET',
  'VERSION',
  '__CTOR_END__',
  '__CTOR_LIST__',
  '__DTOR_END__',
  '__DTOR_LIST__',
]);

const LD_BUILTINS = new Set([
  'ABSOLUTE',
  'ADDR',
  'ALIGN',
  'ALIGNOF',
  'BLOCK',
  'COMMONPAGESIZE',
  'CONSTANT',
  'DATA_SEGMENT_ALIGN',
  'DATA_SEGMENT_END',
  'DATA_SEGMENT_RELRO_END',
  'DEFINED',
  'LENGTH',
  'LOADADDR',
  'LOG2CEIL',
  'MAX',
  'MAXPAGESIZE',
  'MIN',
  'NEXT',
  'ORIGIN',
  'SEGMENT_START',
  'SIZEOF',
  'SIZEOF_HEADERS',
  'l',
  'len',
  'o',
  'org',
  'sizeof_headers',
]);

const LD_WARNINGS = new Set([
  '/DISCARD/',
  ':NONE',
  'EXCLUDE_FILE',
]);

class HighlightLd extends Highlighter {

  static NORMAL = 0;
  static WORD = 1;
  static DQUOTE = 2;
  static DQUOTE_BACKSLASH = 3;
  static SLASH = 4;
  static SLASH_SLASH = 5;
  static SLASH_STAR = 6;
  static SLASH_STAR_STAR = 7;

  constructor(delegate) {
    super(delegate);
    this.word = '';
    this.is_bol = true;
    this.is_cpp = false;
    this.is_cpp_builtin = false;
  }

  feed(input) {
    for (let i = 0; i < input.length; i += this.delta) {
      this.delta = 1;
      let c = input[i];
      switch (this.state) {

      case HighlightLd.NORMAL:
        if (!isascii(c) || isalpha(c) || c == '_' || c == '.' || c == ':') {
          this.state = HighlightLd.WORD;
          this.word += c;
        } else if (c == '/') {
          this.state = HighlightLd.SLASH;
        } else if (c == '"') {
          this.state = HighlightLd.DQUOTE;
          this.push("span", "string");
          this.append(c);
        } else if (c == '#' && this.is_bol) {
          this.is_cpp = true;
          this.push("span", "builtin");
          this.append(c);
        } else if (c == '\n') {
          this.append(c);
          if (this.is_cpp) {
            if (this.is_cpp_builtin) {
              this.pop();
            }
            this.is_cpp = false;
          }
          this.is_cpp_builtin = false;
        } else {
          this.append(c);
        }
        break;

      case HighlightLd.WORD:
        if (!isascii(c) || isalnum(c) || c == '_' || c == '.' || c == '/' || c == ':') {
          this.word += c;
        } else {
          if (this.is_cpp) {
            if (CPP_KEYWORDS.has(this.word)) {
              this.push("span", "builtin");
              this.append(this.word);
              this.pop();
              if (this.is_cpp_builtin) {
                this.pop();
                this.is_cpp_builtin = false;
              }
            } else if (C_CONSTANTS.has(this.word)) {
              this.push("span", "constant");
              this.append(this.word);
              this.pop();
            } else {
              this.append(this.word);
            }
          } else {
            if (LD_KEYWORDS.has(this.word)) {
              this.push("span", "keyword");
              this.append(this.word);
              this.pop();
            } else if (LD_BUILTINS.has(this.word)) {
              this.push("span", "builtin");
              this.append(this.word);
              this.pop();
            } else if (LD_WARNINGS.has(this.word)) {
              this.push("span", "warning");
              this.append(this.word);
              this.pop();
            } else {
              this.append(this.word);
            }
          }
          this.word = '';
          this.epsilon(HighlightLd.NORMAL);
        }
        break;

      case HighlightLd.SLASH:
        if (c == '/') {
          this.push("span", "comment");
          this.append("//");
          this.state = HighlightLd.SLASH_SLASH;
        } else if (c == 'D') {
          // for /DISCARD/ warning keyword
          this.word += "/D";
          this.state = HighlightLd.WORD;
        } else if (c == '*') {
          this.push("span", "comment");
          this.append("/*");
          this.state = HighlightLd.SLASH_STAR;
        } else {
          this.append('/');
          this.epsilon(HighlightLd.NORMAL);
        }
        break;

      case HighlightLd.SLASH_SLASH:
        this.append(c);
        if (c == '\n') {
          this.pop();
          this.state = HighlightLd.NORMAL;
          this.is_cpp = false;
        }
        break;

      case HighlightLd.SLASH_STAR:
        this.append(c);
        if (c == '*')
          this.state = HighlightLd.SLASH_STAR_STAR;
        break;

      case HighlightLd.SLASH_STAR_STAR:
        this.append(c);
        if (c == '/') {
          this.pop();
          this.state = HighlightLd.NORMAL;
        } else if (c != '*') {
          this.state = HighlightLd.SLASH_STAR;
        }
        break;

      case HighlightLd.DQUOTE:
        this.append(c);
        if (c == '"') {
          this.pop();
          this.state = HighlightLd.NORMAL;
        } else if (c == '\\') {
          this.state = HighlightLd.DQUOTE_BACKSLASH;
        }
        break;

      case HighlightLd.DQUOTE_BACKSLASH:
        this.append(c);
        this.state = HighlightLd.DQUOTE;
        break;

      default:
        throw new Error('Invalid state');
      }
      if (this.is_bol) {
        if (!isspace(c))
          this.is_bol = false;
      } else {
        if (c == '\n')
          this.is_bol = true;
      }
    }
  }

  flush() {
    switch (this.state) {
    case HighlightLd.WORD:
      if (this.is_cpp) {
        if (CPP_KEYWORDS.has(this.word)) {
          this.push("span", "builtin");
          this.append(this.word);
          this.pop();
        } else if (C_CONSTANTS.has(this.word)) {
          this.push("span", "constant");
          this.append(this.word);
          this.pop();
        } else {
          this.append(this.word);
          this.pop();
        }
      } else {
        if (LD_KEYWORDS.has(this.word)) {
          this.push("span", "keyword");
          this.append(this.word);
          this.pop();
        } else if (LD_BUILTINS.has(this.word)) {
          this.push("span", "builtin");
          this.append(this.word);
          this.pop();
        } else if (LD_WARNINGS.has(this.word)) {
          this.push("span", "warning");
          this.append(this.word);
          this.pop();
        } else {
          this.append(this.word);
        }
      }
      this.word = '';
      break;
    case HighlightLd.SLASH:
      this.append('/');
      if (this.is_cpp)
        this.pop();
      break;
    case HighlightLd.DQUOTE:
    case HighlightLd.DQUOTE_BACKSLASH:
    case HighlightLd.SLASH_SLASH:
    case HighlightLd.SLASH_STAR:
    case HighlightLd.SLASH_STAR_STAR:
      this.pop();
      break;
    default:
      if (this.is_cpp)
        this.pop();
      break;
    }
    if (this.is_cpp) {
      if (this.is_cpp_builtin) {
        this.pop();
      }
      this.is_cpp = false;
    }
    this.is_cpp_builtin = false;
    this.is_bol = true;
    this.state = HighlightLd.NORMAL;
    this.delegate.flush();
    this.delta = 1;
  }
}

Highlighter.REGISTRY['ld-script'] = HighlightLd;
Highlighter.REGISTRY['lds'] = HighlightLd;
Highlighter.REGISTRY['ld'] = HighlightLd;

File: llamafile/highlight/highlight_ld.cpp
// -*- mode:c++;indent-tabs-mode:nil;c-basic-offset:4;coding:utf-8 -*-
// vi: set et ft=cpp ts=4 sts=4 sw=4 fenc=utf-8 :vi
//
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "highlight.h"

#include <ctype.h>

enum
{
    NORMAL,
    WORD,
    DQUOTE,
    DQUOTE_BACKSLASH,
    SLASH,
    SLASH_SLASH,
    SLASH_STAR,
    SLASH_STAR_STAR,
};

HighlightLd::HighlightLd()
{
}

HighlightLd::~HighlightLd()
{
}

void
HighlightLd::feed(std::string* r, std::string_view input)
{
    int c;
    for (size_t i = 0; i < input.size(); ++i) {
        c = input[i] & 255;
        switch (t_) {

        Normal:
        case NORMAL:
            if (!isascii(c) || isalpha(c) || c == '_' || c == '.' || c == ':') {
                t_ = WORD;
                word_ += c;
            } else if (c == '/') {
                t_ = SLASH;
            } else if (c == '"') {
                t_ = DQUOTE;
                *r += HI_STRING;
                *r += c;
            } else if (c == '#' && is_bol_) {
                is_cpp_ = true;
                *r += HI_BUILTIN;
                *r += c;
            } else if (c == '\n') {
                *r += c;
                if (is_cpp_) {
                    *r += HI_RESET;
                    is_cpp_ = false;
                }
            } else {
                *r += c;
            }
            break;

        Word:
        case WORD:
            if (!isascii(c) || isalnum(c) || c == '_' || c == '.' || c == '/' ||
                c == ':') {
                word_ += c;
            } else {
                if (is_cpp_) {
                    if (is_keyword_cpp(word_.data(), word_.size())) {
                        *r += HI_BUILTIN;
                        *r += word_;
                        *r += HI_RESET;
                    } else if (is_keyword_c_constant(word_.data(),
                                                     word_.size())) {
                        *r += HI_CONSTANT;
                        *r += word_;
                        *r += HI_RESET;
                    } else {
                        *r += word_;
                    }
                } else {
                    if (is_keyword_ld(word_.data(), word_.size())) {
                        *r += HI_KEYWORD;
                        *r += word_;
                        *r += HI_RESET;
                    } else if (is_keyword_ld_builtin(word_.data(),
                                                     word_.size())) {
                        *r += HI_BUILTIN;
                        *r += word_;
                        *r += HI_RESET;
                    } else if (is_keyword_ld_warning(word_.data(),
                                                     word_.size())) {
                        *r += HI_WARNING;
                        *r += word_;
                        *r += HI_RESET;
                    } else {
                        *r += word_;
                    }
                }
                word_.clear();
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case SLASH:
            if (c == '/') {
                *r += HI_COMMENT;
                *r += "//";
                t_ = SLASH_SLASH;
            } else if (c == 'D') {
                // for /DISCARD/ warning keyword
                word_ += "/D";
                t_ = WORD;
            } else if (c == '*') {
                *r += HI_COMMENT;
                *r += "/*";
                t_ = SLASH_STAR;
            } else {
                *r += '/';
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case SLASH_SLASH:
            *r += c;
            if (c == '\n') {
                *r += HI_RESET;
                t_ = NORMAL;
                is_cpp_ = false;
            }
            break;

        case SLASH_STAR:
            *r += c;
            if (c == '*')
                t_ = SLASH_STAR_STAR;
            break;

        case SLASH_STAR_STAR:
            *r += c;
            if (c == '/') {
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c != '*') {
                t_ = SLASH_STAR;
            }
            break;

        case DQUOTE:
            *r += c;
            if (c == '"') {
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c == '\\') {
                t_ = DQUOTE_BACKSLASH;
            }
            break;

        case DQUOTE_BACKSLASH:
            *r += c;
            t_ = DQUOTE;
            break;

        default:
            __builtin_unreachable();
        }
        if (is_bol_) {
            if (!isspace(c))
                is_bol_ = false;
        } else {
            if (c == '\n')
                is_bol_ = true;
        }
    }
}

void
HighlightLd::flush(std::string* r)
{
    switch (t_) {
    case WORD:
        if (is_cpp_) {
            if (is_keyword_cpp(word_.data(), word_.size())) {
                *r += HI_BUILTIN;
                *r += word_;
                *r += HI_RESET;
            } else if (is_keyword_c_constant(word_.data(), word_.size())) {
                *r += HI_CONSTANT;
                *r += word_;
                *r += HI_RESET;
            } else {
                *r += word_;
                *r += HI_RESET;
            }
        } else {
            if (is_keyword_ld(word_.data(), word_.size())) {
                *r += HI_KEYWORD;
                *r += word_;
                *r += HI_RESET;
            } else if (is_keyword_ld_builtin(word_.data(), word_.size())) {
                *r += HI_BUILTIN;
                *r += word_;
                *r += HI_RESET;
            } else if (is_keyword_ld_warning(word_.data(), word_.size())) {
                *r += HI_WARNING;
                *r += word_;
                *r += HI_RESET;
            } else {
                *r += word_;
            }
        }
        word_.clear();
        break;
    case SLASH:
        *r += '/';
        if (is_cpp_)
            *r += HI_RESET;
        break;
    case DQUOTE:
    case DQUOTE_BACKSLASH:
    case SLASH_SLASH:
    case SLASH_STAR:
    case SLASH_STAR_STAR:
        *r += HI_RESET;
        break;
    default:
        if (is_cpp_)
            *r += HI_RESET;
        break;
    }
    is_cpp_ = false;
    is_bol_ = true;
    t_ = NORMAL;
}
